_posts/2024-06-19-optimizing-opensearch-with-fp16-quantization.md

```diff
@@ -0,0 +1,286 @@
+|Force merge segments	|1	|1	|
+|Client instance	|r5.16xlarge	|r5.16xlarge	|
+
+Config ID	|Optimization strategy	|m	|ef_construction	|ef_search	|
+|---	|---	|---	|---	|---	|
+|hnsw1	|Default configuration	|16	|100	|100	|
+|hnsw2	|Balance between latency, memory, and recall	|16	|128	|128	|
+|hnsw3	|Optimize for recall	|16	|256	|256	|
+
+Faiss HNSW SQfp16 requires 4 data nodes---half the number needed for Faiss HNSW (8). This demonstrates that SQfp16 reduces memory requirements by 50%. 
+For more information about estimating the required memory and number of data nodes, see the [Appendix](#appendix-memory-and-data-node-requirement-estimation).
+
+#### Recall and memory results
+
+|Experiment ID	|hnsw-recall@1000	|hnsw-sqfp16-recall@1000	|hnsw memory usage (gb)	|hnsw-sqfp16 memory usage (gb)	|% reduction in memory	|
+|---	|---	|---	|---	|---	|---	|
+|hnsw 1	|0.94	|0.94	|300.28	|157.23	|47.64	|
+|hnsw 2	|0.96	|0.96	|300.28	|157.23	|47.64	|
+|hnsw 3	|0.98	|0.98	|300.28	|157.23	|47.64	|
+
+#### Indexing and query results
+
+|Experiment ID	|hnsw mean throughput (docs/sec)	|hnsw-sqfp16 mean throughput (docs/sec)	|hnsw p90 (ms)	|hnsw-sqfp16 p90 (ms)	|hnsw p99 (ms)	|hnsw-sqfp16 p99 (ms)	|
+|---	|---	|---	|---	|---	|---	|---	|
+|hnsw 1	|7544	|7657	|14.02	|16.99	|19.18	|20.83	|
+|hnsw 2	|7063	|7219	|14.21	|17.44	|18.86	|21.80	|
+|hnsw 3	|6004	|5848	|16.14	|20.85	|17.65	|24.73	|
+
+#### Analysis
+
+* For k=1000, the recall is identical for both Faiss HNSW and Faiss HNSW with SQfp16.
+* Faiss HNSW with SQfp16 requires approximately half the memory as Faiss HNSW (as measured by the required number of data nodes). Based on the [k-NN stats API metrics](https://opensearch.org/docs/latest/search-plugins/knn/api/#stats), the memory usage was reduced by 47.64% by using SQfp16.
+* In most instances, SQfp16 demonstrated better indexing throughput as compared to `fp32` vectors.
+
+## Conclusion
+
+Faiss SQfp16 scalar quantization is a powerful technique that provides significant memory savings while maintaining high recall performance similar to full-precision vectors. Converting vectors to a 16-bit floating-point representation can reduce memory requirements by up to 50%. When combined with SIMD optimization, SQfp16 scalar quantization also enhances indexing throughput and reduces search latency, leading to better overall performance. This method strikes an excellent balance between memory efficiency and accuracy, making it a valuable tool for large-scale similarity search applications.
+
+## Future scope
+
+To achieve even greater memory efficiency, we plan to introduce `int8` quantization support using a [Faiss scalar quantizer](https://github.com/opensearch-project/k-NN/issues/1723) and [Lucene scalar quantizer](https://github.com/opensearch-project/k-NN/issues/1277). 
+This advanced technique will enable a remarkable 75% reduction in memory requirements, or 4x compression, compared to full-precision vectors while maintaining high recall performance. 
```

Comment:
Rather than saying "while maintaining high recall performance", since this hasn't been measured yet, we could say, "and we expect to find minimal reduction in recall" or something.